dkg/nodesigs.go

```diff
@@ -49,33 +62,94 @@ func newNodeSigBcast(operatorAmt int, bcastComp *bcast.Component) nodeSigBcast {
 	return ret
 }
 
+// lockHash returns the lock hash byte array from lockHashCh.
+// Once the data has been received, it stays cached in lockHashData.
+func (n *nodeSigBcast) lockHash(ctx context.Context) ([]byte, error) {
+	n.lhLock.Lock()
+	defer n.lhLock.Unlock()
+
+	if len(n.lockHashData) != 0 {
+		return n.lockHashData, nil
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case n.lockHashData = <-n.lockHashCh:
+		return n.lockHashData, nil
+	}
+}
+
+// allSigs returns true if all the node signatures have been received.
+func (n *nodeSigBcast) allSigs() bool {
+	n.osLock.Lock()
+	defer n.osLock.Unlock()
+
+	for _, sig := range n.sigs {
+		if len(sig) == 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
 // broadcastCallback is the default bcast.Callback for nodeSigBcast.
-func (n *nodeSigBcast) broadcastCallback(_ context.Context, _ peer.ID, _ string, msg proto.Message) error {
-	response, ok := msg.(*dkgpb.MsgNodeSig)
+func (n *nodeSigBcast) broadcastCallback(ctx context.Context, _ peer.ID, _ string, msg proto.Message) error {
+	nodeSig, ok := msg.(*dkgpb.MsgNodeSig)
 	if !ok {
 		return errors.New("invalid node sig type")
 	}
 
-	n.otherSigs <- response
+	sig := nodeSig.GetSignature()
+	msgPeerIdx := int(nodeSig.GetPeerIndex())
+
+	if (msgPeerIdx == n.nodeIdx.PeerIdx) || (msgPeerIdx < 0 || msgPeerIdx >= len(n.peers)) {
+		return errors.New("invalid peer index")
+	}
+
+	peerPubk, err := n.peers[msgPeerIdx].PublicKey()
+	if err != nil {
+		return errors.Wrap(err, "get peer public key")
+	}
+
+	lockHash, err := n.lockHash(ctx)
+	if err != nil {
+		return errors.Wrap(err, "lock hash wait")
+	}
+
+	verified, err := k1util.Verify(peerPubk, lockHash, sig[:len(sig)-1])
+	if err != nil {
+		return errors.Wrap(err, "verify signature")
+	} else if !verified {
+		return errors.New("invalid node signature")
+	}
+
+	n.osLock.Lock()
+	n.sigs[msgPeerIdx] = sig
+	n.osLock.Unlock()
```

Comment:
strongly suggest to stick the pattern of dedicated functions for mutexed data reads/writes: `func (b) setSig`